### PR TITLE
Joomla 4.0: Removing calls to JError::raiseError()

### PR DESCRIPTION
**Pull request by Hackwar**
This PR removes almost all calls to JError::raiseError() and replaces them with either a JViewGenericdataexception or a JUserAuthorizationException.

### CODE REVIEW

File: administrator/components/com_messages/controllers/config.php [hunk 1]
```diff
@@ -37,9 +37,7 @@ public function save()
 
 		if (!$form)
 		{
-			JError::raiseError(500, $model->getError());
-
-			return false;
+			throw new Exception($model->getError(), 500);
```
Comment (wilsonge) [inline administrator/components/com_messages/controllers/config.php]: This one in the controller means we can probably do something better than this?

File: administrator/components/com_users/controllers/groups.php [hunk 2]
```diff
@@ -51,8 +51,7 @@ public function delete()
 	{
 		if (!JFactory::getUser()->authorise('core.admin', $this->option))
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
Comment (wilsonge) [inline administrator/components/com_users/controllers/groups.php]: OK Let's make these 403's whilst we are here

File: administrator/components/com_users/controllers/groups.php [hunk 3]
```diff
@@ -71,8 +70,7 @@ public function publish()
...
```
Comment (wilsonge) [inline administrator/components/com_users/controllers/groups.php]: 403

File: administrator/components/com_users/controllers/groups.php [hunk 4]
```diff
@@ -91,8 +89,7 @@ public function reorder()
...
```
Comment (wilsonge) [inline administrator/components/com_users/controllers/groups.php]: 403

File: administrator/components/com_users/controllers/groups.php [hunk 5]
```diff
@@ -111,8 +108,7 @@ public function saveorder()
...
```
Comment (wilsonge) [inline administrator/components/com_users/controllers/groups.php]: 403

File: administrator/components/com_users/controllers/groups.php [hunk 6]
```diff
@@ -131,8 +127,7 @@ public function checkin()
...
```
Comment (wilsonge) [inline administrator/components/com_users/controllers/groups.php]: 403

File: administrator/components/com_users/controllers/level.php [hunk 7]
```diff
@@ -57,8 +57,7 @@ public function delete()
 
 		if (!JFactory::getUser()->authorise('core.admin', $this->option))
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
Comment (wilsonge) [inline administrator/components/com_users/controllers/level.php]: 403

File: administrator/modules/mod_latest/helper.php [hunk 8]
```diff
@@ -82,9 +82,7 @@ public static function getList(&$params)
 
 		if ($error = $model->getError())
 		{
-			JError::raiseError(500, $error);
-
-			return false;
+			throw new Exception(implode("\n", $errors), 500);
```
Comment (wilsonge) [inline administrator/modules/mod_latest/helper.php]: OK So whilst we are here should a module failing stop the entire page from loading??

File: components/com_banners/models/banner.php [hunk 9]
```diff
@@ -52,7 +52,7 @@ public function click()
 		}
 		catch (RuntimeException $e)
 		{
-			JError::raiseError(500, $e->getMessage());
+			throw new Exception($e->getMessage(), 500);
```
Comment (wilsonge) [inline components/com_banners/models/banner.php]: Can we use `setError` here for now please and `return false` and let the view handle the error for consistency

File: components/com_mailto/controller.php [hunk 10]
```diff
@@ -96,7 +96,7 @@ public function send()
 			{
 				if (strpos($_POST[$field], $header) !== false)
 				{
-					JError::raiseError(403, '');
+					throw new Exception('', 403);
```
Comment (wilsonge) [inline components/com_mailto/controller.php]: Let's define a message for the love of god

File: administrator/modules/mod_latest/helper.php (hunk 8 above)
Comment (mbabker) [inline administrator/modules/mod_latest/helper.php]: This one should probably be caught by the module's base file and enqueue a message I'd say.  Modules probably shouldn't kill the entire page but it'll have to be a case-by-case to figure out what to do.

**Comment by Hackwar** [comment]
I reverted several of the changes to concentrate on the view.*.php files. I also introduced 2 new exceptions, JViewGenericdataexception and JUserAuthorizationException. Those 2 are consistently used throughout the view.*.php files. I'm wondering if it wouldn't be better to create a new branch and new PR with just those view.*.php files...

**Comment by mbabker** [comment]
There are also a couple new Exception classes added for either 3.6.3 or 3.7
from Andre.  I think his was only a controller exception but I say it
because I think it dealt with authorization.

**Comment by Hackwar** [comment]
I've reverted most of the stuff and started over and just concentrated on the view files. Taking babysteps here... I would keep it at this and do further refactoring in other PRs.

File: components/com_mailto/views/mailto/view.html.php [hunk 11]
```diff
@@ -58,9 +58,7 @@ protected function &getData()
 
 		if ($data->link == '')
 		{
-			JError::raiseError(403, JText::_('COM_MAILTO_LINK_IS_MISSING'));
-
-			return false;
+			throw new Exception(JText::_('COM_MAILTO_LINK_IS_MISSING'), 403);
```
Comment (wilsonge) [inline components/com_mailto/views/mailto/view.html.php]: I think this is a 400 not a 403?

**Comment by wilsonge** [comment]
This is beginning to look significantly better :)

**Comment by Hackwar** [comment]
Should we merge this for the moment or do you want more changes before accepting this?

**Pull request merged**